api/src/main/java/org/infinispan/api/sync/StrongCounter.java

```diff
@@ -0,0 +1,98 @@
+package org.infinispan.api.sync;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.infinispan.api.common.events.CounterListener;
+import org.infinispan.api.common.events.ListenerHandle;
+import org.infinispan.api.configuration.CounterConfiguration;
+
+/**
+ * @since 13.0
+ */
+public interface StrongCounter {
+   String name();
+
+
+   /**
+    * It fetches the current value.
+    * <p>
+    * It may go remotely to fetch the current value.
+    *
+    * @return The current value.
+    */
+   CompletableFuture<Long> value();
+
+   /**
+    * Atomically increments the counter and returns the new value.
+    *
+    * @return The new value.
+    */
+   default CompletableFuture<Long> incrementAndGet() {
+      return addAndGet(1L);
+   }
+
+
+   /**
+    * Atomically decrements the counter and returns the new value
+    *
+    * @return The new value.
+    */
+   default CompletableFuture<Long> decrementAndGet() {
+      return addAndGet(-1L);
+   }
+
+
+   /**
+    * Atomically adds the given value and return the new value.
+    *
+    * @param delta The non-zero value to add. It can be negative.
+    * @return The new value.
+    */
+   CompletableFuture<Long> addAndGet(long delta);
+
+   /**
+    * Resets the counter to its initial value.
+    */
+   CompletableFuture<Void> reset();
+
+   /**
+    * Registers a {@link CounterListener} to this counter.
+    *
+    * @param listener The listener to register.
+    * @return A {@link ListenerHandle} that allows to remove the listener via {@link ListenerHandle#remove()}.
+    */
+   ListenerHandle<CounterListener> listen(CounterListener listener);
```

Comment:
Remove the listeners/notifications. I want to drop this when we move the implementation to use JGroups directly instead of caches.

Reply:
So there will no longer be a way to implement threshold listeners ?

Reply:
I would expect users would want some sort of notification when a counter gets to a given value etc.

Reply:
I guess it wouldn't be too difficult to have events in StrongCounter (RAFT is replicated anyway, I can try to find a way to get the deltas from there)

But let's remove events for WeakCounter. I'm planning to implement a PN-Counter in JGroups (https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type#PN-Counter_(Positive-Negative_Counter)). it would be impossible to get any useful event.

Reply:
would it make sense to make the method like "java stream" (I don't want to impl Stream). Examples
`Closeable handle = sCounter.stream().forEach(counterValueAndState -> ...)`
`Closeable handle = sCounter.stream().filterValue(v -> v>6).filterState(s -> s==VALID).forEach(...)`

Reply:
I've removed the listener from the `WeakCounter`
Your "stream" example could be done via reactive publish/subscrib